File: server/kafka-backend/src/application/saga/saga-orchestrator.ts
import { Injectable, Logger } from '@nestjs/common';
import { EntityManager } from 'typeorm';

import { AccountRepository } from '../../core/ports/out/account.repository.interface';
import { TransferRepository } from '../../core/ports/out/transfer.repository.interface';
import { EventPublisher } from '../../core/ports/out/event-publisher.interface';
import { TransferStatus } from '../../core/domain/transfer.entity';
import { RetryService, BackoffStrategy } from '../../infrastructure/services/retry.service';
import type
{
  TransferInitiatedEvent,
  TransferCompletedEvent,
  TransferFailedEvent,
} from '../../core/domain/transfer.entity';

export interface SagaStep
{
  name: string;
  execute: (context: SagaContext) => Promise<void>;
  compensate: (context: SagaContext) => Promise<void>;
  isCritical?: boolean;
}

export interface SagaContext
{
  transferId: string;
  fromAccountId: number;
  toAccountId: number;
  amount: number;
  currency: string;
  reason: string;
  entityManager: EntityManager;
  metadata: Record<string, unknown>;
}

export enum SagaState
{
  NOT_STARTED = 'NOT_STARTED',
  IN_PROGRESS = 'IN_PROGRESS',
  COMPLETED = 'COMPLETED',
  COMPENSATING = 'COMPENSATING',
  COMPENSATED = 'COMPENSATED',
  FAILED = 'FAILED',
}

@Injectable()
export class SagaOrchestrator
{
  private readonly logger = new Logger(SagaOrchestrator.name);
  private readonly sagaStates = new Map<string, SagaState>();

  constructor(
    private readonly accountRepository: AccountRepository,
    private readonly transferRepository: TransferRepository,
    private readonly eventPublisher: EventPublisher,
    private readonly retryService: RetryService,
  ) {}

  async execute(
    event: TransferInitiatedEvent,
    steps: SagaStep[],
  ): Promise<void>
  {
    const { transferId } = event;
    this.sagaStates.set(transferId, SagaState.IN_PROGRESS);

    const context: SagaContext = {
      transferId,
      fromAccountId: event.fromAccountId,
      toAccountId: event.toAccountId,
      amount: event.amount,
      currency: event.currency,
      reason: event.reason,
      entityManager: null as unknown as EntityManager,
      metadata: {},
    };

    const executedSteps: number[] = [];
    let sagaFailed = false;
    let failureReason = '';

    // ---- FORWARD EXECUTION ----
    await this.accountRepository.runInTransaction(
      async (entityManager: EntityManager) =>
      {
        context.entityManager = entityManager;

        for(let i = 0; i < steps.length; i++)
        {
          const step = steps[i];

          try
          {
            this.logger.log(
              `Saga [${transferId}] step ${i + 1}/${steps.length}: ${step.name}`,
            );

            await step.execute(context);
            executedSteps.push(i);
          }
          catch(error)
          {
            const errMsg = error instanceof Error ? error.message : 'Unknown error';
            this.logger.error(`Saga [${transferId}] step ${step.name} failed: ${errMsg}`);

            sagaFailed = true;
            failureReason = `Step "${step.name}" failed: ${errMsg}`;

            // Critical steps are points of no return — continue forward instead of compensating
            if(step.isCritical)
            {
              this.logger.warn(
                `Saga [${transferId}] critical step ${step.name} failed — pushing forward`,
              );
              sagaFailed = false;
              executedSteps.push(i);
              continue;
            }

            throw error;
          }
        }

        if(sagaFailed)
        {
          throw new Error(failureReason);
        }
      },
    );

    // ---- COMPENSATION ----
    if(sagaFailed)
    {
      this.sagaStates.set(transferId, SagaState.COMPENSATING);
      this.logger.warn(
        `Saga [${transferId}] compensating ${executedSteps.length} steps in reverse`,
      );

      await this.compensate(context, steps, executedSteps, failureReason);
      return;
    }

    await this.publishSuccessEvents(context);
    this.sagaStates.set(transferId, SagaState.COMPLETED);
    this.logger.log(`Saga [${transferId}] completed successfully`);
  }

  private async compensate(
    context: SagaContext,
    steps: SagaStep[],
    executedSteps: number[],
    failureReason: string,
  ): Promise<void>
  {
    const reversed = [...executedSteps].reverse();
    let compensationErrors = 0;

    for(const stepIndex of reversed)
    {
      const step = steps[stepIndex];

      try
      {
        await this.retryService.execute(
          () => step.compensate(context),
          {
            maxAttempts: 3,
            baseDelayMs: 500,
            strategy: BackoffStrategy.EXPONENTIAL,
          },
        );

        this.logger.log(`Saga [${context.transferId}] compensated step: ${step.name}`);
      }
      catch(err)
      {
        compensationErrors++;
        this.logger.error(
          { err },
          `Saga [${context.transferId}] compensation failed for step: ${step.name}`,
        );
      }
    }

    await this.transferRepository.updateStatus(
      context.transferId,
      TransferStatus.FAILED,
      undefined,
      failureReason,
    );

    const sagaState = compensationErrors === 0
      ? SagaState.COMPENSATED
      : SagaState.FAILED;
    this.sagaStates.set(context.transferId, sagaState);

    const failedEvent: TransferFailedEvent = {
      eventType: 'transfer.failed',
      transferId: context.transferId,
      fromAccountId: context.fromAccountId,
      toAccountId: context.toAccountId,
      amount: context.amount,
      currency: context.currency,
      failureReason: compensationErrors > 0
        ? `${failureReason} (compensation had ${compensationErrors} errors)`
        : failureReason,
      failedAt: new Date().toISOString(),
    };

    await this.eventPublisher.publish(
      'transfer.failed',
      context.transferId,
      failedEvent as unknown as Record<string, unknown>,
    );
  }

  buildTransferSagaSteps(): SagaStep[]
  {
    return [
      {
        name: 'RESERVE_FUNDS',
        execute: async (ctx: SagaContext) =>
        {
          const sender = await this.accountRepository.findByIdWithLock(
            ctx.fromAccountId,
            ctx.entityManager,
          );

          if(!sender)
          {
            throw new Error(`Sender account ${ctx.fromAccountId} not found`);
          }

          if(sender.moneda !== ctx.currency)
          {
            throw new Error(`Currency mismatch: ${sender.moneda} vs ${ctx.currency}`);
          }

          const balance = Number(sender.sold);
          if(balance < ctx.amount)
          {
            throw new Error(`Insufficient funds: ${balance} < ${ctx.amount}`);
          }

          const newBalance = parseFloat((balance - ctx.amount).toFixed(2));
          await this.accountRepository.updateBalance(
            ctx.fromAccountId,
            newBalance,
            ctx.entityManager,
          );

          ctx.metadata.previousSenderBalance = balance;
        },
        compensate: async (ctx: SagaContext) =>
        {
          const previousBalance = ctx.metadata.previousSenderBalance as number;
          await this.accountRepository.updateBalance(
            ctx.fromAccountId,
            previousBalance,
          );
        },
      },
      {
        name: 'CREDIT_RECEIVER',
        execute: async (ctx: SagaContext) =>
        {
          const receiver = await this.accountRepository.findByIdWithLock(
            ctx.toAccountId,
            ctx.entityManager,
          );

          if(!receiver)
          {
            throw new Error(`Receiver account ${ctx.toAccountId} not found`);
          }

          const newBalance = parseFloat(
            (Number(receiver.sold) + ctx.amount).toFixed(2),
          );

          await this.accountRepository.updateBalance(
            ctx.toAccountId,
            newBalance,
            ctx.entityManager,
          );

          ctx.metadata.previousReceiverBalance = Number(receiver.sold);
        },
        compensate: async (ctx: SagaContext) =>
        {
          const previousBalance = ctx.metadata.previousReceiverBalance as number;
          await this.accountRepository.updateBalance(
            ctx.toAccountId,
            previousBalance,
          );
        },
      },
      {
        name: 'RECORD_TRANSFER',
        isCritical: true,
        execute: async (ctx: SagaContext) =>
        {
          await this.transferRepository.updateStatus(
            ctx.transferId,
            TransferStatus.COMPLETED,
            new Date(),
          );
        },
        compensate: async () => {},
      },
    ];
  }

  private async publishSuccessEvents(context: SagaContext): Promise<void>
  {
    const completedEvent: TransferCompletedEvent = {
      eventType: 'transfer.completed',
      transferId: context.transferId,
      fromAccountId: context.fromAccountId,
      toAccountId: context.toAccountId,
      amount: context.amount,
      currency: context.currency,
      completedAt: new Date().toISOString(),
    };

    await this.eventPublisher.publish(
      'transfer.completed',
      context.transferId,
      completedEvent as unknown as Record<string, unknown>,
    );
  }

  getSagaState(transferId: string): SagaState
  {
    return this.sagaStates.get(transferId) ?? SagaState.NOT_STARTED;
  }
}

File: server/backend/src/main/java/com/intbank/application/saga/SagaOrchestrator.java
package com.intbank.application.saga;

import com.intbank.core.domain.event.TransferCompletedEvent;
import com.intbank.core.domain.event.TransferFailedEvent;
import com.intbank.core.domain.event.TransferInitiatedEvent;
import com.intbank.core.domain.vo.TransferStatus;
import com.intbank.core.port.out.AccountRepository;
import com.intbank.core.port.out.EventPublisher;
import com.intbank.core.port.out.TransferRepository;
import com.intbank.infrastructure.persistence.entity.SagaInstanceJpaEntity;
import com.intbank.infrastructure.persistence.entity.SagaStepJpaEntity;
import com.intbank.infrastructure.persistence.repository.SagaInstanceJpaRepository;
import com.intbank.infrastructure.persistence.repository.SagaStepJpaRepository;
import com.intbank.service.RetryService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
public class SagaOrchestrator
{

    private static final Logger log = LoggerFactory.getLogger(SagaOrchestrator.class);

    private final AccountRepository accountRepository;
    private final TransferRepository transferRepository;
    private final EventPublisher eventPublisher;
    private final RetryService retryService;
    private final SagaInstanceJpaRepository sagaInstanceRepo;
    private final SagaStepJpaRepository sagaStepRepo;

    public SagaOrchestrator(AccountRepository accountRepository, TransferRepository transferRepository,
                            EventPublisher eventPublisher, RetryService retryService,
                            SagaInstanceJpaRepository sagaInstanceRepo, SagaStepJpaRepository sagaStepRepo)
    {
        this.accountRepository = accountRepository;
        this.transferRepository = transferRepository;
        this.eventPublisher = eventPublisher;
        this.retryService = retryService;
        this.sagaInstanceRepo = sagaInstanceRepo;
        this.sagaStepRepo = sagaStepRepo;
    }

    public void execute(TransferInitiatedEvent event, List<SagaStep> steps)
    {
        String transferId = event.trackingId();
        persistState(transferId, SagaState.IN_PROGRESS.name(), null, 0);

        SagaContext context = new SagaContext(
                transferId, event.fromAccountId(), event.toAccountId(),
                event.amount(), event.currency(), event.description(), new HashMap<>());

        List<Integer> executedSteps = new ArrayList<>();
        boolean sagaFailed = false;
        String failureReason = "";

        try
        {
            accountRepository.runInTransaction(() ->
            {
                for (int i = 0; i < steps.size(); i++)
                {
                    SagaStep step = steps.get(i);
                    try
                    {
                        log.info("Saga [{}] step {}/{}: {}", transferId, i + 1, steps.size(), step.name());
                        step.execute().execute(context);
                        markStep(transferId, step.name(), "COMPLETED");
                        executedSteps.add(i);
                    }
                    catch (Exception error)
                    {
                        String errMsg = error.getMessage() != null ? error.getMessage() : "Unknown error";
                        log.error("Saga [{}] step {} failed: {}", transferId, step.name(), errMsg);

                        if (step.isCritical())
                        {
                            log.warn("Saga [{}] critical step {} failed — pushing forward", transferId, step.name());
                            markStep(transferId, step.name(), "FAILED_SKIPPED");
                            executedSteps.add(i);
                            continue;
                        }
                        throw new RuntimeException(errMsg);
                    }
                }
                return null;
            });
        }
        catch (Exception e)
        {
            sagaFailed = true;
            failureReason = e.getMessage() != null ? e.getMessage() : "Saga execution failed";
        }

        if (sagaFailed)
        {
            persistState(transferId, SagaState.COMPENSATING.name(), failureReason, executedSteps.size());
            log.warn("Saga [{}] compensating {} steps in reverse", transferId, executedSteps.size());
            compensate(context, steps, executedSteps, failureReason);
            return;
        }

        publishSuccessEvents(context);
        persistState(transferId, SagaState.COMPLETED.name(), null, executedSteps.size());
        log.info("Saga [{}] completed successfully", transferId);
    }

    private void compensate(SagaContext context, List<SagaStep> steps, List<Integer> executedSteps, String failureReason)
    {
        var reversed = new ArrayList<>(executedSteps);
        Collections.reverse(reversed);
        int compensationErrors = 0;

        for (int stepIndex : reversed)
        {
            SagaStep step = steps.get(stepIndex);
            try
            {
                retryService.execute(() ->
                {
                    step.compensate().compensate(context);
                    return null;
                }, new RetryService.RetryOptions(3, 500, 30000, RetryService.BackoffStrategy.EXPONENTIAL, null));
                log.info("Saga [{}] compensated step: {}", context.transferId(), step.name());
            }
            catch (Exception err)
            {
                compensationErrors++;
                log.error("Saga [{}] compensation failed for step: {}", context.transferId(), step.name(), err);
            }
        }

        transferRepository.updateStatus(context.transferId(), TransferStatus.FAILED, null);

        SagaState finalState = compensationErrors == 0 ? SagaState.COMPENSATED : SagaState.FAILED;
        persistState(context.transferId(), finalState.name(), failureReason, executedSteps.size());

        String finalReason = compensationErrors > 0
                ? failureReason + " (compensation had " + compensationErrors + " errors)"
                : failureReason;

        TransferFailedEvent failedEvent = new TransferFailedEvent(
                context.transferId(), context.fromAccountId(), context.toAccountId(),
                "", "", context.amount(), context.currency(), finalReason, Instant.now());

        Map<String, Object> eventMap = new LinkedHashMap<>();
        eventMap.put("trackingId", failedEvent.trackingId());
        eventMap.put("fromAccountId", failedEvent.fromAccountId());
        eventMap.put("toAccountId", failedEvent.toAccountId());
        eventMap.put("amount", failedEvent.amount());
        eventMap.put("currency", failedEvent.currency());
        eventMap.put("failureReason", failedEvent.failureReason());
        eventMap.put("failedAt", failedEvent.failedAt().toString());

        eventPublisher.publish(TransferFailedEvent.EVENT_NAME, context.transferId(), eventMap);
    }

    public List<SagaStep> buildTransferSagaSteps()
    {
        return List.of(
                new SagaStep("RESERVE_FUNDS",
                        ctx ->
                        {
                            var sender = accountRepository.findByIdWithLock(ctx.fromAccountId())
                                    .orElseThrow(() -> new RuntimeException("Sender account " + ctx.fromAccountId() + " not found"));
                            if (!sender.moneda().equals(ctx.currency()))
                            {
                                throw new RuntimeException("Currency mismatch: " + sender.moneda() + " vs " + ctx.currency());
                            }
                            if (sender.sold().compareTo(ctx.amount()) < 0)
                            {
                                throw new RuntimeException("Insufficient funds: " + sender.sold() + " < " + ctx.amount());
                            }
                            BigDecimal newBalance = sender.sold().subtract(ctx.amount()).setScale(2, java.math.RoundingMode.HALF_EVEN);
                            accountRepository.updateBalance(ctx.fromAccountId(), newBalance);
                            ctx.metadata().put("previousSenderBalance", sender.sold());
                        },
                        ctx ->
                        {
                            BigDecimal previousBalance = (BigDecimal) ctx.metadata().get("previousSenderBalance");
                            accountRepository.updateBalance(ctx.fromAccountId(), previousBalance);
                        },
                        false
                ),
                new SagaStep("CREDIT_RECEIVER",
                        ctx ->
                        {
                            var receiver = accountRepository.findByIdWithLock(ctx.toAccountId())
                                    .orElseThrow(() -> new RuntimeException("Receiver account " + ctx.toAccountId() + " not found"));
                            BigDecimal newBalance = receiver.sold().add(ctx.amount()).setScale(2, java.math.RoundingMode.HALF_EVEN);
                            accountRepository.updateBalance(ctx.toAccountId(), newBalance);
                            ctx.metadata().put("previousReceiverBalance", receiver.sold());
                        },
                        ctx ->
                        {
                            BigDecimal previousBalance = (BigDecimal) ctx.metadata().get("previousReceiverBalance");
                            accountRepository.updateBalance(ctx.toAccountId(), previousBalance);
                        },
                        false
                ),
                new SagaStep("RECORD_TRANSFER",
                        ctx ->
                        {
                            transferRepository.updateStatusWithEntity(ctx.transferId(), TransferStatus.COMPLETED, Instant.now(), null);
                        },
                        ctx ->
                        {
                        },
                        true
                )
        );
    }

    private void publishSuccessEvents(SagaContext context)
    {
        TransferCompletedEvent completedEvent = new TransferCompletedEvent(
                context.transferId(), context.fromAccountId(), context.toAccountId(),
                "", "", context.amount(), context.currency(), Instant.now());

        Map<String, Object> eventMap = new LinkedHashMap<>();
        eventMap.put("trackingId", completedEvent.trackingId());
        eventMap.put("fromAccountId", completedEvent.fromAccountId());
        eventMap.put("toAccountId", completedEvent.toAccountId());
        eventMap.put("amount", completedEvent.amount());
        eventMap.put("currency", completedEvent.currency());
        eventMap.put("completedAt", completedEvent.completedAt().toString());

        eventPublisher.publish(TransferCompletedEvent.EVENT_NAME, context.transferId(), eventMap);
    }

    public SagaState getSagaState(String transferId)
    {
        return sagaInstanceRepo.findById(transferId)
                .map(e -> SagaState.valueOf(e.getState()))
                .orElse(SagaState.NOT_STARTED);
    }

    private void persistState(String transferId, String state, String failureReason, int currentStep)
    {
        SagaInstanceJpaEntity entity = sagaInstanceRepo.findById(transferId)
                .orElseGet(() -> new SagaInstanceJpaEntity());
        entity.setTransferId(transferId);
        entity.setState(state);
        entity.setFailureReason(failureReason);
        entity.setCurrentStep(currentStep);
        sagaInstanceRepo.save(entity);
    }

    private void markStep(String transferId, String stepName, String status)
    {
        SagaStepJpaEntity step = new SagaStepJpaEntity();
        step.setSagaId(transferId);
        step.setStepName(stepName);
        step.setStatus(status);
        step.setExecutedAt(Instant.now());
        sagaStepRepo.save(step);
    }

    public enum SagaState
    {
        NOT_STARTED, IN_PROGRESS, COMPLETED, COMPENSATING, COMPENSATED, FAILED
    }

    public record SagaStep(String name, SagaStepExecutor execute, SagaStepCompensator compensate, boolean isCritical)
    {
        public interface SagaStepExecutor
        {
            void execute(SagaContext ctx) throws Exception;
        }

        public interface SagaStepCompensator
        {
            void compensate(SagaContext ctx) throws Exception;
        }
    }

    public record SagaContext(String transferId, String fromAccountId, String toAccountId, BigDecimal amount,
                              String currency, String reason, Map<String, Object> metadata)
    {
    }
}
